readthedocs/builds/tasks.py

```diff
@@ -453,3 +457,150 @@ def send_build_status(build_pk, commit, status, link_to_build=False):
+            old_webhooks = (
+                self.project.webhook_notifications
+                .filter(events__isnull=True)
+            )
+
+        new_webhooks = (
+            self.project.webhook_notifications
+            .filter(events__name=self.event)
+        )
+        for webhook in itertools.chain(old_webhooks, new_webhooks):
+            try:
+                self.send_webhook(webhook)
+            except Exception:
+                log.exception(
+                    'Failed to send webhook. id=%s',
+                    webhook.id,
+                )
+
+    def send_email(self, email):
+        """Send email notifications for build failure."""
+        # We send only what we need from the Django model objects here to avoid
+        # serialization problems in the ``readthedocs.core.tasks.send_email_task``
+        context = {
+            'version': {
+                'verbose_name': self.version.verbose_name,
+            },
+            'project': {
+                'name': self.project.name,
+            },
+            'build': {
+                'pk': self.build.pk,
+                'error': self.build.error,
+            },
+            'build_url': 'https://{}{}'.format(
+                settings.PRODUCTION_DOMAIN,
+                self.build.get_absolute_url(),
+            ),
+            'unsub_url': 'https://{}{}'.format(
+                settings.PRODUCTION_DOMAIN,
+                reverse('projects_notifications', args=[self.project.slug]),
+            ),
+        }
+
+        if self.build.commit:
+            title = _(
+                'Failed: {project[name]} ({commit})',
+            ).format(commit=self.build.commit[:8], **context)
+        else:
+            title = _('Failed: {project[name]} ({version[verbose_name]})').format(
+                **context
+            )
+
+        log.debug(
+            'Sending email notification. project=%s version=%s build=%s',
+            self.project.slug, self.version.slug, self.build.id,
+        )
+        send_email(
+            email,
+            title,
+            template='projects/email/build_failed.txt',
+            template_html='projects/email/build_failed.html',
+            context=context,
+        )
+
+    def send_webhook(self, webhook):
+        """Send webhook notification for project webhook."""
+        payload = webhook.get_payload(
+            version=self.version,
+            build=self.build,
+            event=self.event,
+        )
+        if not payload:
+            payload = json.dumps({
+                'name': self.project.name,
+                'slug': self.project.slug,
+                'build': {
+                    'id': self.build.id,
+                    'commit': self.build.commit,
+                    'state': self.build.state,
+                    'success': self.build.success,
+                    'date': self.build.date.strftime('%Y-%m-%d %H:%M:%S'),
```

Comment:
I would use isoformat here instead of our custom one, `.isoformat()`.

Reply:
This is the default payload from the old webhooks, we can't change this without breaking those.